request.go

```diff
@@ -73,29 +78,55 @@ func (r *Request) execute(c Client, output interface{}) error {
 	request.Header.Add("Content-Type", "application/json")
 	c.cfg.logger.Debugf("Request body: %v", request.Body)
 
-	//execute the request
-	result, err := c.cfg.HTTPClient.Do(request)
-	if err != nil {
-		return err
-	}
-
-	iostream, err := ioutil.ReadAll(result.Body)
-	if err != nil {
-		return err
-	}
+	timer := time.After(requestServerErrorRetryTimeout)
+	var latestRetryErr error
+RETRY:
+	for {
+		select {
+		case <-timer:
+			c.cfg.logger.Errorf("Timeout reached when retrying, last error: %v", latestRetryErr)
+			return latestRetryErr
+		default:
+			//execute the request
+			result, err := c.cfg.HTTPClient.Do(request)
+			if err != nil {
+				c.cfg.logger.Errorf("Error while executing the request: %v", err)
+				return err
+			}
 
-	c.cfg.logger.Debugf("Status code returned: %v", result.StatusCode)
+			iostream, err := ioutil.ReadAll(result.Body)
+			if err != nil {
+				c.cfg.logger.Errorf("Error while reading the response's body: %v", err)
+				return err
+			}
 
-	if result.StatusCode >= 300 {
-		var errorMessage RequestError //error messages have a different structure, so they are read with a different struct
-		errorMessage.StatusCode = result.StatusCode
-		json.Unmarshal(iostream, &errorMessage)
-		c.cfg.logger.Errorf("Error message: %v. Status: %v. Code: %v.", errorMessage.ErrorMessage, errorMessage.StatusMessage, errorMessage.StatusCode)
-		return errorMessage
+			c.cfg.logger.Debugf("Status code returned: %v", result.StatusCode)
+
+			if result.StatusCode >= 300 {
+				var errorMessage RequestError //error messages have a different structure, so they are read with a different struct
+				errorMessage.StatusCode = result.StatusCode
+				json.Unmarshal(iostream, &errorMessage)
+				if result.StatusCode >= 500 {
+					latestRetryErr = errorMessage
+					c.cfg.logger.Errorf("RETRY! Error message: %v. Title: %v. Code: %v.", errorMessage.Description, errorMessage.Title, errorMessage.StatusCode)
+					time.Sleep(500 * time.Millisecond) //delay the request, so we don't do too many requests to the server
```

Comment:
I would also expose a parameter `maximumNumberOfRetries` compute the waiting delay in between the retries.
`waitingDelay := requestServerErrorRetryTimeout/ maximumNumberOfRetries`

Reply:
I would say either `timeout` or `maxNumberRetries` is chosen for better practice, not both.

Reply:
quote from AWS `In addition to simple retries, each AWS SDK implements exponential backoff algorithm for better flow control. The idea behind exponential backoff is to use progressively longer waits between retries for consecutive error responses. You should implement a maximum delay interval, as well as a maximum number of retries. The maximum delay interval and maximum number of retries are not necessarily fixed values, and should be set based on the operation being performed, as well as other local factors, such as network latency.

Reply:
That is exactly what I stated. Either `timeout` or `maxNumberRetries` is chosen. That means we can use either `timeout` with `delayInterval` or `maxNumberRetries` with `delayInterval`. we should not use `timeout`, `maxNumberRetries` and `delayInterval` at the same time, which is kind of a bad practice.

Reply:
If you want this implementation `waitingDelay := requestServerErrorRetryTimeout/ maximumNumberOfRetries`. We still have to hardcode `maximumNumberOfRetries`.

Reply:
Ok. let's use `maxNumberRetries` with `delayInterval`